Examples/CaseStudies/SwiftUICaseStudies/04-HigherOrderReducers-FailureToast.swift

```diff
@@ -0,0 +1,145 @@
+import Combine
+import ComposableArchitecture
+import SwiftUI
+
+private let readMe = """
+  This screen demonstrates how the `Reducer` struct can be extended to enhance reducers with extra \
+  functionality.
+
+  In it we introduce an `errorHandling` reducer that consumes the failable `appReducer`. \
+  It handles the errors by way of a toast, and returns the usual non-failing reducer.
+
+  This form of reducer is useful if you want to centralize and handle failures in the same way. \
+  Without this, each routine executed in the `appReducer` would need to handle it's own failure.
+
+  Tapping the "Load Data" button will fail after one second and show a toast.
+  """
+
+enum AppError: Error, LocalizedError {
+  case api
+
+  var errorDescription: String? { "Whoops! There was a server error." }
+}
+
+struct AppState: Equatable {
+  var toastStatus: ToastStatus = .hiding
+  var data: [String] = []
+}
+
+enum ToastStatus: Equatable {
+  case showing(String)
+  case hiding
+
+  var isShowing: Bool {
+      switch self {
+      case .showing: return true
+      case .hiding: return false
+      }
+  }
+
+  var text: String {
+      switch self {
+      case .showing(let text): return text
+      case .hiding: return ""
+      }
+  }
+}
+
+enum AppAction {
+  case showToast(String)
+  case hideToast
+  case loadData
+  case didLoadData([String])
+}
+
+struct AppEnvironment {
+  var mainQueue: AnySchedulerOf<DispatchQueue>
+  var loadData: () -> Effect<[String], Error>
+}
+
+let appReducer: (inout AppState, AppAction, AppEnvironment) -> Effect<AppAction, Error> = { state, action, environment in
+  switch action {
+  case .showToast(let text):
+    state.toastStatus = .showing(text)
+    return .none
+
+  case .hideToast:
+    state.toastStatus = .hiding
+    return .none
+
+  case .loadData:
+    return environment.loadData().map(AppAction.didLoadData)
+
+  case .didLoadData(let data):
+    state.data = data
+    return .none
+  }
+}
+
+extension Reducer {
+  static func errorHandling(
+    _ reducer: @escaping (inout AppState, AppAction, AppEnvironment) -> Effect<AppAction, Error>
+  ) -> Reducer<AppState, AppAction, AppEnvironment> {
+    Reducer<AppState, AppAction, AppEnvironment> { state, action, environment in
+      reducer(&state, action, environment)
+        .catch { Just(.showToast($0.localizedDescription)) }
+        .eraseToEffect()
+    }
+  }
```

Comment:
We're aware that this function isn't generic and tied to the `App` component, which is the only component in this example. We're interested in learning how this might work with sub-components as well in a nice way. We've looked at the `resending` technique which helps deal with duplicate actions needed in sub-components. But not sure if that's the right way. Ideally, this could be used on the root component only, and all sub-components would get it automatically also.

Reply:
When I implemented reusable time travel functionality what I did was to have a generic state with what I needed, that nested the rest of the app state, same for the actions. Then the higher order function wraps the app state/action in the resusable functionality state/action. That makes your App feature completely unware of it. The small inconvenience is that you need a custom view that wraps your feature view too. 

If you wanted to show the toast on specific actions that would make it hard (then you would have to make your reusable feature aware of specific actions of your app) but since you're using the `catch` on effect (nice idea!) it should be fine.

Reply:
One of the cases we are hitting is that Reducer return type is `Effect<Action, Never>` and was wondering why we can't have `Reducer<State, Action, Error, Environment>`and return type of `Effect<Action, Error>`? Then we can bubble up the error using pullback.

Reply:
Also, we're looking at another approach that might work. Hoping to have something more concrete soon

Reply:
To clear up the initial comment, we think that making the reducer generic is a separate concern from making this work with sub-components. 

Generics: 
We think we understand in general the technique you describe @alexito4. We looked around for the time travelling example and found a video on your site which we'll have to watch. 🙏 

Sub-components:
[We've experimented with a solution which seems to work](https://github.com/sundeepgupta/swift-composable-architecture/compare/error-toast-case-study...sundeepgupta:error-result-case-study). Essentially, we defined a `FailableReducer` which uses failable `Effect<Action, Error>. This way, the errors from sub-components can bubble up. We're looking to feedback on this approach and how much sense it might make for TCA's `Reducer` to work with failable effects. 🙏

Reply:
Hi all, just wanted to bump this thread to get some thoughts on the failable reducer idea linked above... Any ideas/feedback is welcome. We need this kind of functionality in our app and wondering how good/bad of an idea this might be. Especially, considering potential future changes to TCA that might be on the roadmap